src/database.c

```diff
@@ -156,9 +158,15 @@ static void subhier_clean(struct mosquitto_db *db, struct mosquitto__subhier **s
 		subhier_clean(db, &peer->children);
 		UHPA_FREE_TOPIC(peer);
 
-		HASH_DELETE(hh, *subhier, peer);
-		mosquitto__free(peer);
+		// skip head for preventing to collapse *subhier
```

Comment:
Did you had any issue with deleting head here ? This should works since HASH_DELETE will update *subhier (https://troydhanson.github.io/uthash/userguide.html => Delete all items from a hash).

Reply:
Thanks. I've misunderstood. Reconfirmed the original subhier_clean made same result.